microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/EmittingPublisher.java

```diff
@@ -0,0 +1,170 @@
+                LOGGER.fine("Subscription cancelled");
+                state.compareAndSet(State.NOT_REQUESTED_YET, State.CANCELLED);
+                state.compareAndSet(State.READY_TO_EMIT, State.CANCELLED);
+            }
+
+        });
+    }
+
+    /**
+     * Properly fail the stream, set publisher to cancelled state and send {@code onError} signal downstream.
+     * Signal {@code onError} is sent only once, any other call to this method is no-op.
+     *
+     * @param throwable Sent as {@code onError} signal
+     */
+    void fail(Throwable throwable) {
+        if (!terminated.getAndSet(true) && subscriber != null) {
+            state.compareAndSet(State.NOT_REQUESTED_YET, State.CANCELLED);
+            state.compareAndSet(State.READY_TO_EMIT, State.CANCELLED);
+            this.subscriber.onError(throwable);
+        }
+    }
+
+    /**
+     * Properly complete the stream, set publisher to completed state and send {@code onComplete} signal downstream.
+     * Signal {@code onComplete} is sent only once, any other call to this method is no-op.
+     */
+    void complete() {
+        if (!terminated.getAndSet(true) && subscriber != null) {
+            state.compareAndSet(State.NOT_REQUESTED_YET, State.COMPLETED);
+            state.compareAndSet(State.READY_TO_EMIT, State.COMPLETED);
+            this.subscriber.onComplete();
+        }
+    }
+
+    /**
+     * Emit one item to the stream, if there is enough requested, item is signaled to downstream as {@code onNext}
+     * and method returns true. If there is requested less than 1, nothing is sent and method returns false.
+     *
+     * @param item to be sent downstream
+     * @return true if item successfully sent
+     * @throws java.lang.IllegalStateException if publisher is cancelled
+     */
+    boolean emit(T item) {
+        return this.state.get().emit(this, item);
+    }
+
+    /**
+     * Check if publisher is in terminal state CANCELLED.
+     *
+     * @return true if so
+     */
+    boolean isCancelled() {
+        return this.state.get() == State.CANCELLED;
+    }
+
+    /**
+     * Check if publisher is in terminal state COMPLETED.
+     *
+     * @return true if so
+     */
+    boolean isCompleted() {
+        return this.state.get() == State.COMPLETED;
+    }
+
+    private enum State {
+        NOT_REQUESTED_YET {
+            @Override
+            <T> boolean emit(EmittingPublisher<T> publisher, T item) {
+                return false;
+            }
+        },
+        READY_TO_EMIT {
+            @Override
+            <T> boolean emit(EmittingPublisher<T> publisher, T item) {
+                if (publisher.requested.getAndDecrement() < 1) {
+                    return false;
+                }
+                try {
+                    publisher.subscriber.onNext(item);
+                    return true;
+                } catch (Throwable t) {
+                    publisher.fail(t);
+                    return false;
+                }
```

Comment:
NPE rethrow was intentional, as its reserved as a signal for upstream [(§2.13)](https://github.com/reactive-streams/reactive-streams-jvm#2.13) and this is not the place we are able to solve kafka client errors. On the other hand I am more and more convinced we should remove abstraction layer between `BasicKafkaConsumer`, `BasicKafkaPublisher`,  `EmittingPublisher` and create one specialized `KafkaConsumingPublisher`, which can be used in both Helidon MP and SE